### PR TITLE
[#8] Add plugin builder Dockerfiles

### PR DESCRIPTION
**Pull request by alanking**
Addresses #8
In service of #11

Still testing the platforms:
- [x] centos:7
- [x] almalinux:8
- [x] rockylinux:9
- [x] debian:11
- [x] debian:12
- [x] ubuntu:20.04
- [x] ubuntu:22.04

### CODE REVIEW

**Review by korydraughn — COMMENTED** [review]
Were these files taken from the development environment?

**Comment by korydraughn** [comment]
I agree. Leveraging the dev environment Dockerfiles is the right step IMO.

**Comment by alanking** [comment]
This was not as far along as I'd thought. Converting to draft. Will have this ready to go by COB today, I hope.

**Comment by alanking** [comment]
Okay, got everything working. I went ahead and updated the CMakeLists file as well based on the CMakeLists in the irods/irods repo for plugins/auth.

**Comment by alanking** [comment]
@SwooshyCueb - Feel free to have a look at this. I imagine we'll want to take another pass over this before the release, but this is just to get the ball rolling.

File: CMakeLists.txt [hunk 1]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
+# Testing executable
 add_executable(
   pam_handshake_auth_check
-  src/auth_check_wrapper.cpp
-  src/auth_check_main.cpp
-  src/pam_conversation.cpp
+  "${CMAKE_CURRENT_SOURCE_DIR}/src/auth_check_wrapper.cpp"
+  "${CMAKE_CURRENT_SOURCE_DIR}/src/auth_check_main.cpp"
+  "${CMAKE_CURRENT_SOURCE_DIR}/src/pam_conversation.cpp"
 )
```
Comment (korydraughn) [inline CMakeLists.txt:48]: A separate repo sounds fine.

Or, its inclusion is controlled via a build option.
If it's kept in this repo, no one has to go deal with another repo for testing.

Thoughts?

**Review by korydraughn — COMMENTED** [review]
Looks normal and boring.

Deferring to @SwooshyCueb for further review.

File: CMakeLists.txt (hunk 1 above)
Comment (korydraughn) [reply CMakeLists.txt:48]: Ohhh, got it.

Totally agree with you.

Comment (korydraughn) [reply CMakeLists.txt:48]: Feel free to mark this resolved or leave it open for others to comment on.

File: CMakeLists.txt [hunk 2]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
+set(CPACK_RPM_FILE_NAME RPM-DEFAULT)
 
-# set(CPACK_DEBIAN_PACKAGE_NAME "irods-auth-interactive-pam-server" COMPONENT server)
-set(CPACK_DEBIAN_PACKAGE_DEPENDS "${IRODS_PACKAGE_DEPENDENCIES_STRING},irods-runtime, libcurl4-gnutls-dev")
+set(CPACK_ARCHIVE_COMPONENT_INSTALL OFF)
 
-# set(CPACK_DEBIAN_PACKAGE_NAME "irods-auth-interactive-pam-client" COMPONENT client)
-set(CPACK_DEBIAN_PACKAGE_MAINTAINER "Stefan Wolfsheimer")
+set(CPACK_DEBIAN_PACKAGE_NAME ${PLUGIN_PACKAGE_NAME})
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "${IRODS_PACKAGE_DEPENDENCIES_STRING}, irods-server (= ${IRODS_VERSION}), irods-runtime (= ${IRODS_VERSION}), libc6")
```
Comment (SwooshyCueb) [inline CMakeLists.txt:185]: add `libpam0g` to deb package depends.

File: CMakeLists.txt [hunk 3]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
+
+set(CPACK_RPM_PACKAGE_NAME ${PLUGIN_PACKAGE_NAME})
+set(CPACK_RPM_PACKAGE_REQUIRES "${IRODS_PACKAGE_DEPENDENCIES_STRING}, irods-server = ${IRODS_VERSION}, irods-runtime = ${IRODS_VERSION}")
```
Comment (SwooshyCueb) [inline CMakeLists.txt:188]: add `pam` to rpm package requires.

File: CMakeLists.txt [hunk 4]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
+  "${plugin_target_part}_client"
   PRIVATE
   ${IRODS_COMPILE_DEFINITIONS_PRIVATE}
 )
+install(
+  TARGETS
+  "${plugin_target_part}_client"
+  LIBRARY
+  DESTINATION "${IRODS_PLUGINS_DIRECTORY}/auth"
+  COMPONENT ${IRODS_PACKAGE_COMPONENT_RUNTIME_NAME}
```
Comment (SwooshyCueb) [inline CMakeLists.txt:102]: We're only making one package, so no `COMPONENT` argument is needed.

File: CMakeLists.txt [hunk 5]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
+  pam_handshake_auth_check
+  PRIVATE
+  "${PAM_LIBRARY}"
+)
+install(
+  TARGETS
+  pam_handshake_auth_check
+  RUNTIME
+  DESTINATION "/usr/lib/irods/plugins/auth"
+  COMPONENT server
```
Comment (SwooshyCueb) [inline CMakeLists.txt:59]: We're only making one package, so no `COMPONENT` argument is needed.

File: CMakeLists.txt [hunk 6]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
-install(TARGETS irods_auth_plugin-pam_interactive_client
-        LIBRARY
-        DESTINATION "/usr/lib/irods/plugins/auth/"
-        COMPONENT client
+install(
+  TARGETS
+  "${plugin_target_part}_server"
+  LIBRARY
+  DESTINATION "${IRODS_PLUGINS_DIRECTORY}/auth"
+  COMPONENT ${IRODS_PACKAGE_COMPONENT_RUNTIME_NAME}
```
Comment (SwooshyCueb) [inline CMakeLists.txt:148]: We're only making one package, so no `COMPONENT` argument is needed.

File: docker/builders/plugin_builder.centos7.Dockerfile [hunk 7]
```diff
@@ -1,39 +1,24 @@
...
-    rm -rf /var/cache/yum /tmp/*
+RUN --mount=type=cache,target=/var/cache/yum,sharing=locked \
+    yum update -y || [ "$?" -eq 100 ] && \
+    rm -rf /tmp/*
 
-RUN yum install -y \
+RUN --mount=type=cache,target=/var/cache/yum,sharing=locked \
+    yum install -y \
         epel-release \
+        gcc-c++ \
```
Comment (SwooshyCueb) [inline docker/builders/plugin_builder.centos7.Dockerfile:15]: You'll want `centos-release-scl` as well. Our new clang externals package will depend on packages from the SCL repos.

**Review by SwooshyCueb — CHANGES_REQUESTED** [review]
Just a few notes

File: CMakeLists.txt (hunk 4 above)
Comment (alanking) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 5 above)
Comment (alanking) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 6 above)
Comment (alanking) [reply CMakeLists.txt:146]: Done

File: docker/builders/plugin_builder.centos7.Dockerfile (hunk 7 above)
Comment (alanking) [reply docker/builders/plugin_builder.centos7.Dockerfile:16]: Done

File: CMakeLists.txt [hunk 8]
```diff
@@ -58,132 +38,158 @@ if (NOT PAM_LIBRARY)
...
-        LIBRARY
-        DESTINATION "/usr/lib/irods/plugins/auth/"
-        COMPONENT server
-)
+include(IrodsCPackCommon)
+
+list(APPEND CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "${CPACK_PACKAGING_INSTALL_PREFIX}${IRODS_HOME_DIRECTORY}")
+list(APPEND CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "${CPACK_PACKAGING_INSTALL_PREFIX}${IRODS_HOME_DIRECTORY}/scripts")
+list(APPEND CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "${CPACK_PACKAGING_INSTALL_PREFIX}${IRODS_HOME_DIRECTORY}/scripts/irods")
+list(APPEND CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "${CPACK_PACKAGING_INSTALL_PREFIX}${IRODS_HOME_DIRECTORY}/scripts/irods/test")
```
[the remark this answers is not in the document]
Comment (alanking) [reply CMakeLists.txt:156]: We will be adding tests for #6 and they will go into the usual spot for tests, so I think we should just leave these in. Thoughts?

File: CMakeLists.txt (hunk 1 above)
Comment (alanking) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 5 above)
Comment (alanking) [reply CMakeLists.txt]: Oh, good catch. Done

File: CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply CMakeLists.txt:155]: Makes sense.

**Comment by korydraughn** [comment]
Assuming #6 will be addressed in a different PR, is there anything left to do for this PR?

**Comment by alanking** [comment]
I don't think so. I was able to build packages for all the platforms and they look kinda like what we want, at least for now, so that's the goal for this one.

**Comment by korydraughn** [comment]
Cool. If you're happy with it, squash to taste.

**Comment by alanking** [comment]
Squashed

**Review by korydraughn — APPROVED** [review]
Let's get an approval from @SwooshyCueb before pounding.

**Review by SwooshyCueb — APPROVED** [review]
LGTM

**Pull request merged**